kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,268 @@
+    fn read(
+        &self,
+        region: &Region<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Write data from a buffer to flash.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// starting at the chosen `start_block`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `range` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT prepare the block for writing first.
+    /// The user of this function SHOULD ensure that a block is erased
+    /// before calling `write` (see `erase`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it SHOULD NOT be written again until it's erased as part of an erase block.
+    /// Multiple writes to the same block are possible in this trait,
+    /// but the result is dependent on the nature of the underlying hardware,
+    /// and this interface does not provide a way to discover that.
+    ///
+    /// **Note** about raw flash devices: writes can turn bits from `1` to `0`.
+    /// To change a bit from `0` to `1`, a region must be erased.
+    ///
+    /// **Note**: some raw flash hardware only allows a limited number of writes before
+    /// an erase. If that is the case, the implementation MUST return an error
+    /// `ErrorCode::NOMEM` when this happens, even if the hardware silently
+    /// ignores the write.
+    ///
+    /// On success, triggers `Client::write_complete` once.
+    /// On failure returns a `ErrorCode` and the buffer passed in.
+    fn write(
+        &self,
+        region: &Region<W>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Makes a region ready for writing.
+    ///
+    /// This corresponds roughly to the erase operation on raw flash devices,
+    /// but may also do nothing on devices where erasure is not necessary.
+    ///
+    /// Calling `erase` may modify bytes in the selected `region` in any way
+    /// (typically, on raw flash devices, all bits will be set,
+    /// i.e. each byte turns to 0xFF).
+    ///
+    /// If `region` exceeds the size of the device, returns `ErrorCode::INVAL`.
+    /// On success, triggers `Client::erase_complete` once.
+    fn erase(&self, region: &Region<E>) -> Result<(), ErrorCode>;
+
+    /// Returns the size of the device in bytes.
+    fn get_size(&self) -> Result<u64, ErrorCode>;
+}
+
+pub trait HasClient<'a, C> {
+    /// Set the client for this peripheral. The client will be called
+    /// when operations complete.
+    fn set_client(&'a self, client: &'a C);
+}
+
+/// Implement `Client` to receive callbacks from `BlockStorage`.
+pub trait Client<const W: u32, const E: u32> {
+    /// Block read complete.
+    ///
+    /// This will be called when the read operation is complete.
+    fn read_complete(&self, read_buffer: &'static mut [u8], ret: Result<(), ErrorCode>);
```

Comment:
What are the errors?

Reply:
I left it for now. I presume that any virtualization needs to store the read offset anyway, so storing read size alongside it wouldn't be a problem. Or am I mistaken?

Reply:
Given both BlockIndex and AddressRange as borrowed with an anonymous lifetime, I think this is OK. It does necessitate that the caller keeps that state. I see benefits and drawbacks in both approaches (having the caller keep the state, or having the callee pass it back).